rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+specification, `SendChannel` is given first class support with
+`"type": "sendchannel"` and `value` set to the UUID of the
+channel. This enables an important pattern: to receive messages from a
+remote context, you can send it a `SendChannel` object to use for
+responses.
+
+For deserialization, primitive values are converted back to
+primitives, but complex values are represented by a `RemoteObject`
+type. In cases like arrays where there is a `value` field holding a
+container object, `RemoteObject.toLocal()` recursively converts the
+content of the container into local objects (so e.g. a `type: array` is
+converted into a local `Array` instances, and any contained `type: object`
+objects are converted into local `Object` instances, and so on through
+the full tree).
+
+#### Higher Level API
+
+The low-level API provides required primitives, but it's difficult to
+use directly. To achieve the aim of making tests no harder to write
+than SpecialPowers-based Gecko tests, there is also a higher-level API
+initially providing two main capabilities: the ability to
+`postMessage` a remote context, and the ability to `executeScript` so
+that the script runs in a remote context.
+
+This API is provided by a `RemoteWindow` object. The `RemoteWindow`
+object doesn't handle creating the browsing context, but given the
+UUID for the remote window, creates a `SendChannel` which is able to
+send messages to the remote. Alternatively the `RemoteWindow` may be
+created first and its `uuid` property used when constructing the URL.
+
+Inside the remote browsing context itself, the test author has to call
+`await start_window()` in order to set up a `RecvChannel` with UUID given
+by the `uuid` parameter in `location.href`. The returned object offers
+an `addMessageHandler(callback)` API to receive messages sent with the
+`postMessage` API on `RemoteWindow`, and `async nextMessage()` to wait
+for the next message.
+
+##### Script Execution
+
+`RemoteWindow.executeScript(fn, ...args)` serializes the function
+`fn`, using `Function.toString()`. Each argument is
+serialized using the remote value serialization algorithm. Along with
+the function string and the arguments, a `SendChannel` is sent for the
+command response (only one such channel is created per `RemoteWindow`
+for efficiency reasons), and a command id is sent to disambiguate
+responses from different commands.
+
+On the remote side the function is deserialized and executed. If
+execution results in a `Promise` value, the result of that promise is
+awaited. The final return value after the promise is resolved is sent
+back and forms the async return value of the `executeScript` call. If
+the script throws, the thrown value is provided as the result, and
+re-thrown in the originating context. In addition an `exceptionDetails`
+field provides the line/column numbers of the original exception,
+where available.
+
+In addition there is a `RemoteWindow.executeScriptNoResult(fn,
```

Comment:
Based on my experience on bfcache tests, I feel `executeScriptNoResult()` might be hard to use, because:

- Mixing `await executeScriptNoResult()` and `await executeScript()` might complicates the timeline. Having `await executeScript()` only would be clear, because it serializes the script evaluation on caller and remote context.
- The choice between `await executeScriptNoResult()` and `await executeScript()` might depend on internals of `executeScript()`. When I was migrating my bfcache tests from COOP-based framework to test_driver-based framework, I had to make changes around `executeScriptNoResult()`-equivalent that were hard to explain without knowing the internal structure of the framework, so I switched to the `executeScript()`-only interface.

So I'd prefer removing `executeScriptNoResult()`.
Still there would remain complexities around navigation: instead of `executeScriptNoResult()`, I created promises / async callbacks triggered by `executeScript()` but not blocking `async executeScript()` in bfcache tests. But by removing `executeScriptNoResult()` I expect we can make the complexities only navigation-specific, not affecting `RemoteWindow` interface.

Reply:
So (for the context of others) the technical concern here is that for bfcache tests in particular you need all sockets to be closed before initiating the navigation. The problem is that if you send a message, execute some script that closes all sockets, and then send the result, you end up re-opening a socket for the result. So there are broadly two choices:
* Don't try to send a result at all. `postMessage` works like this, and `executeScriptNoResult` does the same: we simply don't provide a channel to send the result on so there's no concern about sockets being reopened.
* Send a result, but ensure we navigate after the result has been sent, so nothing will implictly reopen a socket. Practically this means something like `setTimeout(async url => {await closeAllSockets(); startNavigation(url)}, 0)`. The problem with that is that we are currently linting for `setTimeout` so the infra discourages even usage that's required like this.

I don't quite get the argument about serialization; in the case that you're putting the actual action of the remote script in a timeout the test window can guarantee the script has been sent, but the actual action of the script hasn't happened at the time the value is returned, so it doesn't seem like that method offers particuarly stronger guarantees around ordering. I suppose one advantage is that any exception that happens when evalutaing the script is reported through the return value, rather than just happening in the remote window.

So, I think I'm mostly OK with dropping `executeScriptNoResult` because, as you say, it's exposing an implementation detail and it seems cleaner to start with a smaller API. My concern is that for some kinds of tests (notably bfcache), understanding that implementation detail is very important, and there's an argument that making it an explicit part of the API will help people write correct tests. So I'd like to hear other opinions here.